src/simplify_algebra.cpp

```diff
@@ -412,24 +412,83 @@ struct find_dot_mul
     }
 };
 
+
+
+auto fusable_split()
+{
+    return match::make_basic_pred_matcher([](instruction_ref ins) {
+
+        std::vector<instruction_ref> slices;
+
+        for (const auto& output : ins->outputs())
+        {
+            if (output->name() == "slice")
+            {
+                slices.push_back(output);
+            }
+        }
+
+        if (slices.empty())
+            return false;
+
+        std::vector<instruction_ref> add_instructions;
+        for (const auto& slice : slices)
+        {
+            bool used_by_add = false;
+            for (auto& user : slice->outputs())
+            {
+                if (user->name() == "add")
+                {
+                    used_by_add = true;
+                    add_instructions.push_back(user);
+                }
+            }
+            if (!used_by_add)
+                return false;
+        }        
+
+        bool any_add_followed_by_mul = false;
+        for (const auto& add_ins : add_instructions)
+        {
+            for (auto& user : add_ins->outputs())
+            {
+                if (user->name() == "mul")
+                {
+                    any_add_followed_by_mul = true;
+                    ins->outputs();
+                }
+            }
+        }
+
+        if (!any_add_followed_by_mul)
+            return false;
+
+        return true;
+    });
+}
+
+
 // ******************************
 //  a * (x + b) => a * x + a * b
 // ******************************
 // When a * (x + b) is followed by another add of constant, then the
 // additional add can be const folded. Also, better fusions can be applied
 // when the add comes after.
+
 struct find_mul_add
 {
     auto matcher() const
     {
         return match::name("mul")(match::either_arg(0, 1)(
-            match::name("add")(
-                match::either_arg(0, 1)(
-                    match::any().bind("x"),
-                    match::any_of(conv_const_weights(), match::is_constant()).bind("b")),
-                match::none_of(match::args(match::is_constant(), match::is_constant())),
-                match::used_once()),
-            match::is_constant().bind("a")));
+                match::name("add")(
+                    match::either_arg(0, 1)(
+                        match::none_of(match::name("slice")(match::arg(0)(
+                            fusable_split().bind("slc")
+                        ))).bind("x"),
+                        match::any_of(conv_const_weights(), match::is_constant()).bind("b")),
+                    match::none_of(match::args(match::is_constant(), match::is_constant())),
+                    match::used_once()),
+                match::is_constant().bind("a")));
```

Comment:
Break up the matcher with variables to be more readable

Reply:
@CharlieL7 I thought we had a cppcheck for this. I wonder why it doesnt trigger.

Reply:
It probably only checks that there are no more than 3 sets of parenthesis on a single line?